cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,122 +1265,716 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
-  uint8_t** h_dst_bufs = reinterpret_cast<uint8_t**>(h_src_and_dst_buffers.data() + src_bufs_size);
-  // device-side
-  rmm::device_buffer d_src_and_dst_buffers(src_bufs_size + dst_bufs_size + offset_stack_size,
-                                           stream,
-                                           rmm::mr::get_current_device_resource());
-  auto const** d_src_bufs = reinterpret_cast<uint8_t const**>(d_src_and_dst_buffers.data());
-  uint8_t** d_dst_bufs    = reinterpret_cast<uint8_t**>(
-    reinterpret_cast<uint8_t*>(d_src_and_dst_buffers.data()) + src_bufs_size);
-
-  // setup src buffers
-  setup_src_buf_data(input.begin(), input.end(), h_src_bufs);
-
-  // setup dst buffers
-  std::transform(out_buffers.begin(), out_buffers.end(), h_dst_bufs, [](auto& buf) {
-    return static_cast<uint8_t*>(buf.data());
-  });
+  return partition_buf_size_and_dst_buf_info;
+}
+
+/**
+ * @brief Struct containing information about the actual batches we will send to the
+ * `copy_partitions` kernel and the number of iterations we need to carry out this copy.
+ *
+ * For the non-chunked contiguous_split case, this contains the batched dst_buf_infos and the
+ * number of iterations are going to be 1, since the non-chunked case is single pass.
+ *
+ * For the chunked_pack case, this also contains the batched dst_buf_infos for all
+ * iterations in addition to helping keep the state about what batches have been copied so far
+ * and what are the sizes (in bytes) of each iteration.
+ */
+struct chunk_iteration_state {
+  chunk_iteration_state(rmm::device_uvector<dst_buf_info> _d_batched_dst_buf_info,
+                        rmm::device_uvector<offset_type> _d_batch_offsets,
+                        std::vector<std::size_t>&& _h_num_buffs_per_iteration,
+                        std::vector<std::size_t>&& _h_size_of_buffs_per_iteration,
+                        std::size_t total_size)
+    : num_iterations(_h_num_buffs_per_iteration.size()),
+      current_iteration(0),
+      starting_batch(0),
+      d_batched_dst_buf_info(std::move(_d_batched_dst_buf_info)),
+      d_batch_offsets(std::move(_d_batch_offsets)),
+      h_num_buffs_per_iteration(std::move(_h_num_buffs_per_iteration)),
+      h_size_of_buffs_per_iteration(std::move(_h_size_of_buffs_per_iteration)),
+      total_size(total_size)
+  {
+  }
 
-  // HtoD src and dest buffers
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    d_src_bufs, h_src_bufs, src_bufs_size + dst_bufs_size, cudaMemcpyDefault, stream.value()));
+  /**
+   * @brief As of the time of the call, return the starting 1MB batch index, and the
+   * number of batches to copy.
+   *
+   * @return the current iteration's starting_batch and batch count as a pair
+   */
+  std::pair<std::size_t, std::size_t> get_current_starting_index_and_buff_count() const
+  {
+    CUDF_EXPECTS(current_iteration < num_iterations,
+                 "current_iteration cannot exceed num_iterations");
```

Comment:
Can an external caller ever trigger this exception, or would this be purely an internal error? If the latter, maybe not something to throw an exception for, but more like an assert?

Reply:
Hmm I guess this could happen if someone tried to advance the iteration on the state object without first checking if it has any chunks left?

Reply:
Yeap, that's what it is protecting against.